lib/Dialect/FIRRTL/InnerSymbolTable.cpp

```diff
@@ -0,0 +1,152 @@
+//===- InnerSymbolTable.cpp - InnerSymbolTable and InnerRef verification --===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements InnerSymbolTable and verification for InnerRef's.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/FIRRTL/FIRRTLOpInterfaces.h"
+#include "mlir/IR/Threading.h"
+
+using namespace circt;
+using namespace firrtl;
+
+namespace circt {
+namespace firrtl {
+
+//===----------------------------------------------------------------------===//
+// InnerSymbolTable
+//===----------------------------------------------------------------------===//
+
+InnerSymbolTable::InnerSymbolTable(Operation *op) {
+  assert(op->hasTrait<OpTrait::InnerSymbolTable>() &&
+         "expected operation to have InnerSymbolTable trait");
+  assert(op->getNumRegions() == 1 &&
+         "expected operation to have a single region");
+  assert(llvm::hasSingleElement(op->getRegion(0)) &&
+         "expected operation to have a single block");
+
+  // Save
+  this->innerSymTblOp = op;
+
+  // Build table
+
+  // Add all operations
```

Comment:
In general (i.e. elsewhere) comments should end in a period and we usually reflow them in paragraphs.  In particular this looks a bit unfinished